pkg/odo/cli/init/init.go

```diff
@@ -0,0 +1,306 @@
+package init
+
+import (
+	"fmt"
+	"net/url"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/library/pkg/devfile/parser"
+	"github.com/devfile/library/pkg/devfile/parser/data/v2/common"
+	registryLibrary "github.com/devfile/registry-support/registry-library/library"
+
+	"github.com/redhat-developer/odo/pkg/catalog"
+	"github.com/redhat-developer/odo/pkg/component"
+	"github.com/redhat-developer/odo/pkg/devfile"
+	"github.com/redhat-developer/odo/pkg/log"
+	"github.com/redhat-developer/odo/pkg/odo/cli/init/asker"
+	"github.com/redhat-developer/odo/pkg/odo/cli/init/params"
+	"github.com/redhat-developer/odo/pkg/odo/cmdline"
+	"github.com/redhat-developer/odo/pkg/odo/genericclioptions"
+	odoodoutil "github.com/redhat-developer/odo/pkg/odo/util"
+	odoutil "github.com/redhat-developer/odo/pkg/odo/util"
+	"github.com/redhat-developer/odo/pkg/preference"
+	"github.com/redhat-developer/odo/pkg/segment"
+	"github.com/redhat-developer/odo/pkg/testingutil/filesystem"
+	"github.com/redhat-developer/odo/pkg/util"
+
+	"k8s.io/kubectl/pkg/util/templates"
+)
+
+// RecommendedCommandName is the recommended command name
+const RecommendedCommandName = "init"
+
+var initExample = templates.Examples(`
+  # Boostrap a new project in interactive mode
+  %[1]s
+`)
+
+type InitOptions struct {
+	// Backends to build init parameters
+	backends []params.ParamsBuilder
+
+	// filesystem on which command is running
+	fsys filesystem.Filesystem
+
+	// Clients
+	preferenceClient preference.Client
+
+	// the parameters needed to run the init procedure
+	params.InitParams
+
+	// Destination directory
+	destDir string
+}
+
+// NewInitOptions creates a new InitOptions instance
+func NewInitOptions(backends []params.ParamsBuilder, fsys filesystem.Filesystem, prefClient preference.Client) *InitOptions {
+	return &InitOptions{
+		backends:         backends,
+		fsys:             fsys,
+		preferenceClient: prefClient,
+		destDir:          ".",
+	}
+}
+
+// Complete will build the parameters for init, using different backends based on the flags set,
+// either by using flags or interactively is no flag is passed
+// Complete will return an error immediately if the current working directory is not empty
+func (o *InitOptions) Complete(cmdline cmdline.Cmdline, args []string) (err error) {
+
+	empty, err := isEmpty(o.fsys, o.destDir)
+	if err != nil {
+		return err
+	}
+	if !empty {
+		return errors.New("The current directory is not empty. You can bootstrap new component only in empty directory.\nIf you have existing code that you want to deploy use `odo deploy` or use `odo dev` command to quickly iterate on your component.")
+	}
+
+	flags := cmdline.GetFlags()
+	done := false
+	for _, backend := range o.backends {
+		if backend.IsAdequate(flags) {
+			o.InitParams, err = backend.ParamsBuild()
+			if err != nil {
+				return err
+			}
+			done = true
+			break
```

Comment:
Why this? The way two backends' `IsAdequate` method is implemented, it's an either-or situation, IIUC. Unless I'm missing something, we don't need this `break`.

Reply:
This piece of code gets the first backend that is adequate. The fact that the 2 backends are complementary is not considered in this part, because we could want to change conditions in the future in this complementarity could disappear

Reply:
Makes sense. Thanks for explaining. 👍🏾